pkg/test/framework/components/istio/discovery.go

```diff
@@ -0,0 +1,325 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package istio
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net"
+	"sync"
+	"time"
+
+	envoyconfigcorev3 "github.com/envoyproxy/go-control-plane/envoy/config/core/v3"
+	envoyservicediscoveryv3 "github.com/envoyproxy/go-control-plane/envoy/service/discovery/v3"
+	"github.com/hashicorp/go-multierror"
+	"google.golang.org/grpc"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	kube2 "istio.io/istio/pkg/kube"
+	"istio.io/istio/pkg/test"
+	"istio.io/istio/pkg/test/framework/resource"
+	"istio.io/istio/pkg/test/kube"
+	"istio.io/istio/pkg/test/scopes"
+)
+
+// Discovery allows interacting with the discovery server in a cluster to check for configuration correctness.
+type Discovery interface {
+	CallDiscovery(req *envoyservicediscoveryv3.DiscoveryRequest) (*envoyservicediscoveryv3.DiscoveryResponse, error)
+	CallDiscoveryOrFail(t test.Failer, req *envoyservicediscoveryv3.DiscoveryRequest) *envoyservicediscoveryv3.DiscoveryResponse
+
+	StartDiscovery(req *envoyservicediscoveryv3.DiscoveryRequest) error
+	StartDiscoveryOrFail(t test.Failer, req *envoyservicediscoveryv3.DiscoveryRequest)
+	WatchDiscovery(duration time.Duration, accept func(*envoyservicediscoveryv3.DiscoveryResponse) (bool, error)) error
+	WatchDiscoveryOrFail(t test.Failer, duration time.Duration, accept func(*envoyservicediscoveryv3.DiscoveryResponse) (bool, error))
+}
+
+// Discovery returns the service discovery client for the given cluster or the first configured
+// cluster in the environment if given nil.
+func (i *operatorComponent) Discovery(cluster resource.Cluster) (Discovery, error) {
+	cluster = i.ctx.Clusters().GetOrDefault(cluster)
+
+	// lazy init since it requires some k8s calls, port-forwarding and building grpc clients
+	if len(i.discovery) == 0 {
+		if err := i.initDiscovery(); err != nil {
+			return nil, err
+		}
+	}
+
+	if d := i.discovery[cluster.Index()]; d != nil {
+		return d, nil
+	}
+
+	return nil, fmt.Errorf("no service discovery client for %s", cluster.Name())
+}
+
+// DiscoveryOrFail returns the service discovery client for the given cluster or the first configured
+// cluster in the environment if given nil. The test will be failed if there is an error.
+func (i *operatorComponent) DiscoveryOrFail(f test.Failer, cluster resource.Cluster) Discovery {
```

Comment:
Maybe we should just get rid of the cluster on this interface for now? AFAICT none of the tests care (pass in nil). WDYT?